syncnotification.go

```diff
@@ -273,208 +289,268 @@ func (lw *LibWallet) updateAddressDiscoveryProgress() {
 	return int32(math.Ceil(estimatedHeadersDifference))
 }
 
-func (lw *LibWallet) notifySyncError(code SyncErrorCode, err error) {
-	lw.syncData.syncing = false
-	lw.syncData.synced = false
-	lw.activeSyncData = nil // to be reintialized on next sync
+func (mw *MultiWallet) notifySyncError(code SyncErrorCode, err error) {
+	mw.resetSyncData()
 
-	for _, syncProgressListener := range lw.syncData.syncProgressListeners {
+	mw.syncData.mu.RLock()
+	defer mw.syncData.mu.RUnlock()
+
+	for _, syncProgressListener := range mw.syncData.syncProgressListeners {
 		syncProgressListener.OnSyncEndedWithError(err)
 	}
 }
 
-func (lw *LibWallet) notifySyncCanceled() {
-	lw.syncData.syncing = false
-	lw.syncData.synced = false
-	lw.activeSyncData = nil // to be reintialized on next sync
+func (mw *MultiWallet) notifySyncCanceled() {
+	mw.resetSyncData()
+
+	mw.syncData.mu.RLock()
+	defer mw.syncData.mu.RUnlock()
+
+	for _, syncProgressListener := range mw.syncData.syncProgressListeners {
+		syncProgressListener.OnSyncCanceled(mw.syncData.restartSyncRequested)
+	}
+}
+
+func (mw *MultiWallet) resetSyncData() {
+	mw.syncData.mu.Lock()
+	defer mw.syncData.mu.Unlock()
+
+	mw.syncData.syncing = false
+	mw.syncData.synced = false
+	mw.syncData.activeSyncData = nil // to be reintialized on next sync
 
-	for _, syncProgressListener := range lw.syncData.syncProgressListeners {
-		syncProgressListener.OnSyncCanceled(lw.syncData.restartSyncRequested)
+	for _, wallet := range mw.wallets {
+		wallet.waiting = true
 	}
 }
 
-func (lw *LibWallet) synced(synced bool) {
-	lw.syncData.syncing = false
-	lw.syncData.synced = true
-	lw.activeSyncData = nil // to be reintialized on next sync
+func (mw *MultiWallet) synced(walletID int, synced bool) {
+	mw.syncData.mu.Lock()
+	defer mw.syncData.mu.Unlock()
+	if (mw.syncData.synced == synced) && synced {
+		return
+	}
+
+	wallet := mw.wallets[walletID]
+	wallet.synced = synced
+	wallet.syncing = false
+	if mw.OpenedWalletsCount() == mw.SyncedWalletsCount() {
+		mw.syncData.syncing = false
+		mw.syncData.synced = true
+		mw.syncData.activeSyncData = nil // to be reintialized on next sync
 
-	// begin indexing transactions after sync is completed,
-	// syncProgressListeners.OnSynced() will be invoked after transactions are indexed
-	lw.IndexTransactions(func() {
-		for _, syncProgressListener := range lw.syncProgressListeners {
+		for _, syncProgressListener := range mw.syncData.syncProgressListeners {
 			if synced {
 				syncProgressListener.OnSyncCompleted()
 			} else {
 				syncProgressListener.OnSyncCanceled(false)
 			}
 		}
-	})
+
+		// begin indexing transactions after sync is completed,
+		// syncProgressListeners.OnSynced() will be invoked after transactions are indexed
+		var waitForIndexing sync.WaitGroup
+		waitForIndexing.Add(len(mw.wallets))
+		for _, wallet := range mw.wallets {
+			go wallet.IndexTransactions(&waitForIndexing)
+		}
+
+		go func() {
+			waitForIndexing.Wait()
+			mw.syncData.mu.Lock()
```

Comment:
We don't need a lock/unlock to access the listeners. `mw.IsSynced()` performs the only needed read lock and unlock.